src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2242,28 +2249,120 @@ private TypeWithState InferResultNullability(BinaryOperatorKind operatorKind, Me
 
-            if (binary.OperatorKind.IsUserDefined() && binary.MethodOpt?.ParameterCount == 2)
+            var method = binary.MethodOpt;
+
+            if (binary.OperatorKind.IsUserDefined() &&
+                method?.ParameterCount == 2)
             {
+                // Update method based on inferred operand type.
+                TypeSymbol methodContainer = method.ContainingType;
+                bool isLifted = binary.OperatorKind.IsLifted();
+                TypeWithState leftUnderlyingType = GetNullableUnderlyingTypeIfNecessary(isLifted, leftType);
+                TypeWithState rightUnderlyingType = GetNullableUnderlyingTypeIfNecessary(isLifted, rightType);
+                TypeSymbol asMemberOfType = isOrDerivedFrom(methodContainer, leftUnderlyingType.Type) ?
+                    leftUnderlyingType.Type :
+                    (isOrDerivedFrom(methodContainer, rightUnderlyingType.Type) ? rightUnderlyingType.Type : null);
+                if (!(asMemberOfType is null))
+                {
+                    method = (MethodSymbol)AsMemberOfType(asMemberOfType, method);
+                }
+
                 // Analyze operator call properly (honoring [Disallow|Allow|Maybe|NotNull] attribute annotations) https://github.com/dotnet/roslyn/issues/32671
-                var parameters = binary.MethodOpt.Parameters;
-                ReportArgumentWarnings(binary.Left, leftType, parameters[0]);
-                ReportArgumentWarnings(binary.Right, rightType, parameters[1]);
+                var parameters = method.Parameters;
+                visitOperandConversion(binary.Left, leftOperand, leftConversion, parameters[0], leftUnderlyingType);
+                visitOperandConversion(binary.Right, rightOperand, rightConversion, parameters[1], rightUnderlyingType);
+
+                void visitOperandConversion(
+                    BoundExpression expr,
+                    BoundExpression operand,
+                    Conversion conversion,
+                    ParameterSymbol parameter,
+                    TypeWithState operandType)
+                {
+                    _ = VisitConversion(
+                        expr as BoundConversion,
+                        operand,
+                        conversion,
+                        parameter.TypeWithAnnotations,
+                        operandType,
+                        checkConversion: true,
+                        fromExplicitCast: false,
+                        useLegacyWarnings: false,
+                        AssignmentKind.Argument,
+                        parameter);
+                }
+            }
+            else
+            {
+                // Assume this is a built-in operator in which case the parameter types are unannotated.
+                visitOperandConversion(binary.Left, leftOperand, leftConversion, leftType);
+                visitOperandConversion(binary.Right, rightOperand, rightConversion, rightType);
+
+                void visitOperandConversion(
```

Comment:
>visitOperandConversion [](start = 21, length = 22)

nit: I find local functions in the middle of a method. Would you mind moving to the end so it doesn't interrupt the flow of code? Thanks